skills/contextual-embeddings/guide.ipynb

```diff
@@ -0,0 +1,1372 @@
+    "    \n",
+    "    average_score = total_score / total_queries\n",
+    "    pass_at_n = average_score * 100\n",
+    "    return {\n",
+    "        \"pass_at_n\": pass_at_n,\n",
+    "        \"average_score\": average_score,\n",
+    "        \"total_queries\": total_queries\n",
+    "    }\n",
+    "\n",
+    "def evaluate_db_advanced(db, original_jsonl_path, k):\n",
+    "    original_data = load_jsonl(original_jsonl_path)\n",
+    "    \n",
+    "    def retrieval_function(query, db, k):\n",
+    "        return retrieve_rerank(query, db, k)\n",
+    "    \n",
+    "    results = evaluate_retrieval_rerank(original_data, retrieval_function, db, k)\n",
+    "    print(f\"Pass@{k}: {results['pass_at_n']:.2f}%\")\n",
+    "    print(f\"Average Score: {results['average_score']}\")\n",
+    "    print(f\"Total queries: {results['total_queries']}\")\n",
+    "    return results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 380,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [01:22<00:00,  2.99it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@5: 91.24%\n",
+      "Average Score: 0.912442396313364\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [01:34<00:00,  2.63it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@10: 94.79%\n",
+      "Average Score: 0.9479166666666667\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [02:08<00:00,  1.93it/s]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@20: 96.30%\n",
+      "Average Score: 0.9630376344086022\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "results5 = evaluate_db_advanced(contextual_db, 'data/evaluation_set.jsonl', 5)\n",
+    "results10 = evaluate_db_advanced(contextual_db, 'data/evaluation_set.jsonl', 10)\n",
+    "results20 = evaluate_db_advanced(contextual_db, 'data/evaluation_set.jsonl', 20)"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "py311",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.11.6"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```

Comment:
might be worth having a short blurb at the end of what the takeaways are (in bullets) and the most relevant use cases outside of just coding.